sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableClassShouldBeSealed.cs

```diff
@@ -0,0 +1,116 @@
+        private const string MessageFormat = "Make this class 'sealed' or implement 'IEqualityComparer<T>' instead.";
+        private const string EqualsMethodName = nameof(object.Equals);
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                    if (classSymbol == null ||
+                        classSymbol.IsSealed ||
+                        classSymbol.IsStatic ||
+                        classDeclaration.Identifier.IsMissing)
+                    {
+                        return;
+                    }
+
+                    if (HasInvalidCombination(classSymbol))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation()));
+                    }
+
+                }, SyntaxKind.ClassDeclaration);
+        }
+
+        private static bool IsValidEquatableInterfaceSymbol(INamedTypeSymbol namedTypeSymbol)
+        {
+            return namedTypeSymbol.ConstructedFrom.Is(KnownType.System_IEquatable_T) &&
+                namedTypeSymbol.TypeArguments.Length == 1;
+        }
+
+        private static bool IsValidEqualsMethodSymbol(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol.MethodKind == MethodKind.Ordinary &&
+                methodSymbol.Name == EqualsMethodName &&
+                !methodSymbol.IsOverride &&
+                methodSymbol.ReturnType.Is(KnownType.System_Boolean) &&
+                methodSymbol.Parameters.Length == 1;
+        }
+
+        private static bool HasInvalidCombination(INamedTypeSymbol classSymbol)
+        {
+            var equatableInterfacesByTypeName = classSymbol.AllInterfaces
+                .Where(IsValidEquatableInterfaceSymbol)
+                .ToDictionary(nts => nts.TypeArguments[0].Name, nts => nts);
+            var equalsMethodsByTypeName = classSymbol.GetMembers(EqualsMethodName)
+                .OfType<IMethodSymbol>()
+                .Where(IsValidEqualsMethodSymbol)
+                .ToDictionary(ms => ms.Parameters[0].Type.Name, ms => ms);
+
+            // Checks whether any IEquatable<T> has no implementation OR a non-virtual implementation
+            foreach (var iequatable in equatableInterfacesByTypeName)
+            {
+                var associatedMethod = equalsMethodsByTypeName.GetValueOrDefault(iequatable.Key);
+                if (associatedMethod == null ||
```

Comment:
If both conditions are short, I'd move them to one line.

It's a minor style comment, feel free to ignore.